lib/ansible/modules/cloud/docker/docker_swarm_service.py

```diff
@@ -750,6 +750,26 @@ def compare(self, os):
             force_update = True
         return not differences.empty or force_update, differences, needs_rebuild, force_update
 
+    def has_publish_changed(self, old_publish):
+        if len(self.publish) != len(old_publish):
+            return True
+        publish_sorter = operator.itemgetter('published_port', 'target_port', 'protocol')
+        publish = sorted(self.publish, key=publish_sorter)
+        old_publish = sorted(old_publish, key=publish_sorter)
```

Comment:
Sorting will die if any one of the entries of the two lists does not contain all of `published_port`, `target_port` or `protocol`. Is it guaranteed that docker will always return all three elements in each item?

Reply:
To make sure that we always have all three values defined, we can modify the `argument_spec` definition:
```
        publish=dict(default=[], type='list', elements='dict', options=dict(
            published_port=dict(type='int', required=True),
            target_port=dict(type='int', required=True),
            protocol=dict(type='str', required=True),
            mode=dict(type='str', required=False),
        )),
```
This will also prevent people adding other fields (Ansible will complain) and make sure the entries have the correct type.

The documentation can also be improved, by adding `suboptions:` to it. See for example docker_container.py for examples on how this works.

Reply:
Yes as far as I can read from the [docker-py](https://github.com/docker/docker-py/blob/master/docker/types/services.py#L500) code and the [docker api documentation](https://docs.docker.com/engine/api/v1.25/#operation/ServiceCreate) `types.EndpointSpec(ports=ports)` are used. 

Ah cool, I'll add that. `protocol` will not be required by the user as it defaults to `tcp`.

Reply:
I guess forcing the user to be more explicit should be OK, too. Defaults can always be problematic, remember the default user `root` ;-)

Reply:
Yes that makes sense too. I'll leave `protocol` as required then.